### PR TITLE
Implements a socket.io client extension, adds a test case and updates th...

### PR DESCRIPTION
**Pull request by himynameschris**
...e android makefiles.

Documentation on use can be found in the extension header file SocketIO.h or in the test case in the TestCpp project.

Contains the following tasks:
-initial socket.io extension commit
-sioclientimpl subclasses websocket:delegate to respond to websocket events
-implement static connect methods and basic client creation
-move SocketIO class files into correct extension folder (network)
-create SocketIO test in TestCpp
-update project references
-add missing static modifier to connect method
-implement basic test methods
-update extensions tests with SocketIO test entry
-implement basic handshake and opensocket methods for SocketIO
-add Delegate class to handle callbacks, implement virtual Delegate methods in test
-implement socket and client registries for lookup when connecting to hosts and endpoints
-connect delegate onOpen method by separating impl creation from connection
-update test to demonstrate onOpen callback
-create send and emit methods, move SIOClient into header file, add send and emit methods to test
-implement basic socket.io message parsing
-improve logging for events and messages
-add logic to pull event name from payload
-schedule heartbeat to keep connection alive, scheduled for 90% of the heartbeat interval from the server for safety
-add onConnect handler to to catch socket.io onconnect vs websocket onopen
-add disconnect and disconnectFromEndpoint methods to properly disconnect and destroy objects
-modify SIOClientImpl to track _uri for easier lookup in registries
-connect handler for onMessage to message event from socket.io, modify onError handler to take a string instead of WebSocket error code
-create SIOEvent callback type, implement event registry in clients, add test for event registration and callback
-update SIOEvent to use std::function and c++11, utilize cocos2d CC_CALLBACK method to bind selectors, this ensures that the *this reference is properly passed
-check for connect before sending or emitting in the client, cleanup some codes
-change connect logic to reuse existing socket connections instead of opening a new one
-implements get and set Tag methods for clients for easy reference
-improve endpoint handling, add endpoint tests to test layer
-additional error handling within socket disconnect error and failure to open connection
-fixes extracting endpoint from socket.io messages (in cases of the connect message, where there is 1 less colon for the connect message to the default namespace). Also fixes connecting to the default namespace "/" in the connectToEndpoint method
-add disconnect and onClose handlers to client so that onClose is called in the delegate
-add disconnect test methods to test layers
-change c-style casts to static_casts when using a CCDICT_FOREACH
-remove some unneeded namespace completion
-add usage documentation
-add handling for disconnect from server, cleanup some codes
-update comments and documentation in the socketiotest
-update includes so the NDK doesn't complain when compiling with c++11
-add socketio.cpp and test.cpp to the android makefiles
-update test URL to my public server, test script can also be found in my repo at https://github.com/hannon235/socket.io-testserver.git

Signed-off-by: Chris Hannon himynameschris@gmail.com

### CODE REVIEW

**Comment by dumganhar** [comment]
Looks good. I'll test it. Thanks.

File: extensions/network/SocketIO.h [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+     *  @brief  Static client creation method, similar to socketio.connect(uri) in JS
+     *  @param  delegate The delegate which want to receive events from the socket.io client
+     *  @param  uri      The URI of the socket.io server
+     *  @return An initialized SIOClient if connected successfully, otherwise NULL
+     */
+	static SIOClient* connect(SocketIO::SIODelegate& delegate, const std::string& uri);
+	
+	SIOClientImpl* getSocket(const std::string& uri);
+	void addSocket(const std::string& uri, SIOClientImpl* socket);
+	void removeSocket(const std::string& uri);
```
Comment (dumganhar) [inline extensions/network/SocketIO.h]: How about making these three methods as `private` ?

Comment (dumganhar) [inline extensions/network/SocketIO.h]: Probably you need to mark `SIOClientImpl;` as the friend class of `SocketIO`.
Are there something I misunderstood?

**Comment by himynameschris** [comment]
Good catch, I've moved those methods and others into private and used friend to take care of the Impl

**Comment by dumganhar** [comment]
@hannon235 
I have sent you a Pull Request https://github.com/hannon235/cocos2d-x/pull/1 .
Please merge it.
Added socketio files to iOS and Mac projects, and sync with upstream.

**Comment by himynameschris** [comment]
merged, thanks!

**Comment by dumganhar** [comment]
Merged, thanks.

**Pull request merged**